### PR TITLE
Updated report controller

### PR DESCRIPTION
**Pull request by skanthed**
Cron job for the day, week, and month time period.

### CODE REVIEW

**Comment by skanthed** [comment]
There are multiple commits now because olm-bundle was giving an error due to the Operator-SDK version. If everything looks good, I will squash all the commits into one before PR gets merged.

File: main.go [hunk 1]
```diff
@@ -39,6 +36,8 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/healthz"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
 
+	"github.com/go-logr/logr"
+	"github.com/jackc/pgx/v4"
```
Comment (timflannagan) [inline main.go:40]: nit: third party package imports should go in the second grouping.

**Review by timflannagan — COMMENTED** [review]
I think we're getting closer with simplifying the Report controller but I'm hesitant with pushing forward a Report API surface that leaks quite a bit of the underlying details we should aim to abstract:

- Can we continue to surface a spec.Schedule where a user can provide a cron expression? If we want to provide the functionality to produce daily, weekly, monthly, etc. Reports, then we can document the requisite cron expression that users can supply.
- Exposing the spec.CronJobNamespace field in the Report API can result in users mis-configuring that namespace given there's no guarantee we the proper RBAC has been setup correctly. Can we simply hardcode the koku-metrics-operator namespace for now?
- I'm definitely hesitant about surfacing the database-related fields directly into the Report API as that leads to a leaky abstraction, and a source of tech debt down the line.

File: controllers/report_controller.go [hunk 2]
```diff
@@ -48,79 +44,132 @@ type ReportReconciler struct {
...
-	nextRunTime := nextReportPeriod.PeriodEnd
-	waitTime := nextRunTime.Sub(now)
+	// TODO(tflannag): Support updating an existing CronJob resource?
+	l.Info("cronjob already exists -- not creating another one")
+	return nil
+}
 
-	if err := r.Status().Update(ctx, report); err != nil {
-		return ctrl.Result{}, err
+func newCronJobFromReport(d *curatorv1alpha1.Report, scheme *runtime.Scheme) *batchv1.CronJob {
```
[the remark this answers is not in the document]
Comment (skanthed) [reply controllers/report_controller.go:92]: I was using it for checking what I am getting in fetch data and will remove it from here.

File: controllers/report_controller.go [hunk 3]
```diff
@@ -48,79 +44,132 @@ type ReportReconciler struct {
...
+		TypeMeta: metav1.TypeMeta{
+			Kind: "Cronjob",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      d.Name,
+			Namespace: d.Spec.CronjobNamespace,
+		},
+		Spec: batchv1.CronJobSpec{
+			ConcurrencyPolicy: "Forbid",
+			Schedule:          scheduleOfReport,
```
[the remark this answers is not in the document]
Comment (skanthed) [reply controllers/report_controller.go:120]: Sure

File: main.go (hunk 1 above)
Comment (skanthed) [reply main.go:40]: This was added by Ajinkya. I will let him know and confirm this.

Comment (skanthed) [reply main.go:40]: Resolved

**Review by ajinkyababar — APPROVED** [review]
lgtm

**Comment by AvRajath** [comment]
Tested the code on my machine. LGTM.

**Review by AvRajath — APPROVED** [review]

**Pull request merged**